ev3/legoev3_motor.c

```diff
@@ -882,365 +737,79 @@ static enum hrtimer_restart legoev3_motor_timer_callback(struct hrtimer *timer)
-
-		case STATE_SETUP_RAMP_REGULATION:
-			ev3_tm->ramp.count    = 0;
-
-			ev3_tm->state = STATE_RAMP_UP;
-			reprocess = true;
-			break;
+		rampdown_time = (abs(ev3_tm->speed) * ev3_tm->tm.active_params.ramp_down_sp) / (1 + ev3_tm->info->max_speed);
 
-		/*
-		 * The LIMITED_XXX functions have to handle the three phases (any of
-		 * which are optional) of a motor move operation. It is assumed that
-		 * when the run mode was set, the ramp factors were calculated.
-		 *
-		 * The LIMITED_XXX functions need to handle the following combinations:
-		 *
-		 * REGULATED_TIME    - Speed is regulated, ramping is time based
-		 * REGULATED_TACHO   - Speed is regulated, ramping is tacho based
-		 * UNREGULATED_TIME  - Speed is not regulated, ramping is time based
-		 * UNREGULATED_TACHO - Speed is not regulated, ramping is tacho based
-		 *
-		 * When ramping, the code needs to figure out which combination is in
-		 * use, and that's handled by a couple of booleans in the motor struct.
-		 *
-		 * Regardless of the direction of the ramp (up or down), the first part
-		 * of the sequence is ramping up, the tail end of the sequence is
-		 * ramping down.
-		 */
+		rampdown_endpoint = ev3_tm->position
+				  + ((ev3_tm->speed * rampdown_time) / (2*1000) );
+			
+		// TODO: Clean up this target reaching code - STOP happens too soon!
 
-		case STATE_RAMP_UP:
-			/*
-			 * Figure out if we're done ramping up - if yes set state to RAMP_CONST
-			 * and allow states to get reprocessed
-			 */
-			if (IS_POS_CMD(ev3_tm->run_command)) {
-				adjust_ramp_for_position(ev3_tm);
+		if (ev3_tm->speed_reg_sp > 0) {
+			if (rampdown_endpoint > ev3_tm->position_sp) {
+				ev3_tm->speed_reg_sp = (2*1000 * (ev3_tm->position_sp - ev3_tm->position) / (1 + rampdown_time));
 			}
 
-			if (ev3_tm->ramp.count >= ev3_tm->ramp.up.end) {
-				ev3_tm->state = STATE_RAMP_CONST;
-				reprocess = true;
+			if (ev3_tm->position >= ev3_tm->position_sp) {
+				schedule_work(&ev3_tm->notify_position_ramp_down_work);
+				ev3_tm->run_command = TM_COMMAND_STOP;
 			}
-
-			/* Figure out how far along we are in the ramp operation */
-
-			ev3_tm->ramp.percent = calculate_ramp_progress( ev3_tm->ramp.count, ev3_tm->ramp.up.full);
-
-			update_motor_speed_or_power(ev3_tm, ev3_tm->ramp.percent);
-			break;
-
-		case STATE_RAMP_CONST:
-			/*
-			 * Figure out if we're done with the const section - if yes set state to RAMP_DOWN
-			 * and allow states to get reprocessed.
-			 *
-			 *  Just push out the end point if we're in TM_RUN_FOREVER mode
-			 */
-
-			if (ev3_tm->run_command == TM_COMMAND_RUN_FOREVER) {
-				ev3_tm->ramp.down.start = ev3_tm->ramp.count;
-				ev3_tm->ramp.down.end   = ev3_tm->ramp.count + ((abs(ev3_tm->active_params.duty_cycle_sp) * ev3_tm->active_params.ramp_down_sp) / 100);
+		} else if (ev3_tm->speed_reg_sp < 0) {
+			if (rampdown_endpoint < ev3_tm->position_sp) {
+				ev3_tm->speed_reg_sp = (2000 * (ev3_tm->position_sp - ev3_tm->position) / (1 + rampdown_time));
```

Comment:
Inconsistent use of `2*1000` and `2000). Maybe make this mathematical expression a macro since it is used multiple times for readability and maintainability.